docs/sphinx/examples/cpp/providers/photonics_tbi_get_state.cpp

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+// Compile and run with:
+// ```
+// nvq++ --target photonics-cpu photonics_tbi_get_state.cpp -o
+// photonics_tbi_get_state.x && ./photonics_tbi_get_state.x
+// ```
+
+#include "cudaq.h"
+#include "cudaq/photonics.h"
+
+#include <iostream>
+
+// Global variables
+static const std::size_t one{1};
+
+static constexpr std::size_t n_modes{4};
+static constexpr std::array<std::size_t, n_modes> input_state{2, 1, 3, 1};
+
+static constexpr std::size_t d{
+    std::accumulate(input_state.begin(), input_state.end(), one)};
+
+struct TBI {
+  auto operator()(std::vector<double> const &bs_angles,
+                  std::vector<double> const &ps_angles,
+                  std::vector<std::size_t> const &input_state,
+                  std::vector<std::size_t> const &loop_lengths) __qpu__ {
+    auto n_modes = ::n_modes;
+    const auto d = ::d;
+
+    cudaq::qvector<d> qumodes(n_modes); // |00...00> d-dimensions
+    for (std::size_t i = 0; i < n_modes; i++) {
+      for (std::size_t j = 0; j < input_state[i]; j++) {
+        plus(qumodes[i]); // setting to |input_state>
+      }
+    }
+
+    std::size_t c = 0;
+    for (std::size_t ll : loop_lengths) {
+      for (std::size_t i = 0; i < (n_modes - ll); i++) {
+        beam_splitter(qumodes[i], qumodes[i + ll], bs_angles[c]);
+        phase_shift(qumodes[i], ps_angles[c]);
+        c++;
+      }
+    }
+  }
+};
+
+template <typename T>
+void LinearSpacedArray(std::vector<T> &xs, T min, T max, std::size_t N) {
+  T h = (max - min) / static_cast<T>(N - 1);
+  typename std::vector<T>::iterator x;
+  T val;
+  for (x = xs.begin(), val = min; x != xs.end(); ++x, val += h) {
```

Comment:
```suggestion
  
  T val;
  for (auto x = xs.begin(), val = min; x != xs.end(); ++x, val += h) {
```
`x` may not be the clearest name, since it is a quantum gate name.

Reply:
Need to use `typename std::vector<T>::iterator x;`
rename xs to array and x to array_it